core/src/main/java/com/datastrato/gravitino/lock/TreeLockNode.java

```diff
@@ -0,0 +1,145 @@
+
+  private final AtomicLong referenceCount = new AtomicLong();
+
+  public TreeLockNode(NameIdentifier identifier) {
+    this.ident = identifier;
+    this.readWriteLock = new ReentrantReadWriteLock();
+    this.childMap = new ConcurrentHashMap<>();
+  }
+
+  public NameIdentifier getIdent() {
+    return ident;
+  }
+
+  // Why is this method synchronized, please see the comment in the method
+  // LockManager#evictStaleNodes
+  public synchronized void addReference() {
+    referenceCount.getAndIncrement();
+  }
+
+  public void decReference() {
+    referenceCount.getAndDecrement();
+  }
+
+  public long getReferenceCount() {
+    return referenceCount.get();
+  }
+
+  /**
+   * Try to lock the node with the given lock type.
+   *
+   * @param lockType The lock type to lock the node.
+   * @return True if the node is locked successfully, false otherwise.
+   */
+  public boolean tryLock(LockType lockType) {
+    boolean success =
+        lockType == LockType.READ
+            ? readWriteLock.readLock().tryLock()
+            : readWriteLock.writeLock().tryLock();
+    if (success) {
+      referenceCount.getAndIncrement();
+
+      LOG.info(
+          "Node {} has been lock with '{}' lock, reference count = {}",
+          ident,
+          lockType,
+          referenceCount.get());
+    }
+
+    return success;
+  }
+
+  /**
+   * Lock the node with the given lock type. This method should be followed by {@link
+   * #unlock(LockType)}.
+   *
+   * @param lockType The lock type to lock the node.
+   */
+  public void lock(LockType lockType) {
+    if (lockType == LockType.READ) {
+      readWriteLock.readLock().lock();
+    } else {
+      readWriteLock.writeLock().lock();
+    }
+
+    LOG.info(
+        "Node {} has been lock with '{}' lock, reference count = {}",
+        ident,
+        lockType,
+        referenceCount.get());
+  }
+
+  /**
+   * Unlock the node with the given lock type. This method should be called after {@link
+   * #lock(LockType)}, and the lock type should be the same as the lock type in {@link
+   * #lock(LockType)}.
+   *
+   * @param lockType The lock type to unlock the node.
+   */
+  public void unlock(LockType lockType) {
+    if (lockType == LockType.READ) {
+      readWriteLock.readLock().unlock();
+    } else {
+      readWriteLock.writeLock().unlock();
+    }
+    LOG.info(
+        "Node {} has been unlock with '{}' lock, reference count = {}",
+        ident,
+        lockType,
+        referenceCount.get());
+  }
+
+  /**
+   * Get the tree lock node by the given name identifier. If the node doesn't exist, create a new
+   * TreeNode.
+   *
+   * @param ident The name identifier of a resource such as entity or others.
+   * @return The tree lock node of this ident.
+   */
+  public TreeLockNode getOrCreateChild(NameIdentifier ident) {
+    TreeLockNode treeNode = childMap.get(ident);
+    if (treeNode == null) {
+      synchronized (this) {
```

Comment:
Is it enough to use this node level lock? Say for example, if you're acquiring a lock to create/get nodes, your parent get a lock to delete children, what will be happened? I was think we should have a tree level lock for tree modification.

Reply:
Let me think, a global lock seems too large.